torch/distributed/fsdp/_init_utils.py

```diff
@@ -75,13 +88,140 @@
 @no_type_check
 def _init_process_group_state(
     state: _FSDPState,
-    process_group: Optional[dist.ProcessGroup],
+    process_group: ProcessGroupType,
+    sharding_strategy: ShardingStrategy,
+    policy: Optional[_FSDPPolicy],
 ) -> _FSDPState:
-    state.process_group = process_group or _get_default_group()
+    if sharding_strategy in HYBRID_SHARDING_STRATEGIES:
+        if process_group is None and policy is None:
+            # Raise an error here, since this is manual wrapping with no process group
+            # passed in, there is no way to ensure all wrapped FSDP units use the same
+            # process groups.
+            raise ValueError(
+                f"Manual wrapping with {sharding_strategy} requires explicit specification of process group."
+            )
+        else:
+            state = _init_process_group_state_for_hybrid_shard(state, process_group)
+            assert state.process_group is not None, "Expected to populate state.process_group for hybrid shard"
+            assert state._inter_node_pg is not None, "Expected to populate state._inter_node_pg for hybrid shard"
+            assert state._inter_node_state is not None, "Expected to populate state._inter_node_state for hybrid shad."
+    else:
+        if process_group is not None and not isinstance(process_group, dist.ProcessGroup):
+            raise ValueError(
+                f"process_group should be None or dist.ProcessGroup, but got {type(process_group)}"
+            )
+        if process_group is None:
+            state.process_group = _get_default_group()
+        else:
+            state.process_group = process_group
+
     state.rank = state.process_group.rank()
     state.world_size = state.process_group.size()
+
     return state
 
+@no_type_check
+def _init_process_group_state_for_hybrid_shard(state, process_group) -> _FSDPState:
```

Comment:
Is it worth type annotating `process_group` here? Even adding `Any` may be valuable to show that we need to validate the typing in this function.

Reply:
[Easy] Let us type annotate this actually